txnkv/transaction/txn.go

```diff
@@ -791,30 +1063,93 @@ func (txn *KVTxn) LockKeys(ctx context.Context, lockCtx *tikv.LockCtx, keysInput
+		// is too old and CheckExistence is not supported.
+		keyStr := string(key)
+		var val tikv.ReturnedValue
+		var ok bool
+		if val, ok = lockCtx.Values[keyStr]; ok {
+			// TODO: Check if it's safe to use `val.Exists` instead of assuming empty value.
+			if lockCtx.ReturnValues || checkedExistence || val.LockedWithConflictTS != 0 {
 				if !val.Exists {
-					valExists = tikv.SetKeyLockedValueNotExists
+					valExists = false // tikv.SetKeyLockedValueNotExists
 				}
 			}
 		}
-
-		if lockCtx.LockOnlyIfExists && valExists == tikv.SetKeyLockedValueNotExists {
-			skipedLockKeys++
-			continue
+		//}
+		if txn.aggressiveLockingContext != nil {
+			txn.aggressiveLockingContext.currentLockedKeys[keyStr] = tempLockBufferEntry{
+				HasReturnValue:    lockCtx.ReturnValues,
+				HasCheckExistence: lockCtx.CheckExistence,
+				Value:             val,
+			}
+			txn.aggressiveLockingDirty = true
+		} else {
+			setValExists := tikv.SetKeyLockedValueExists
+			if !valExists {
+				setValExists = tikv.SetKeyLockedValueNotExists
+			}
+			// TODO: Fix the calculation when aggressive-locking is active
+			if lockCtx.LockOnlyIfExists && !valExists {
+				skipedLockKeys++
+				continue
+			}
+			memBuf.UpdateFlags(key, tikv.SetKeyLocked, tikv.DelNeedCheckExists, setValExists)
 		}
-		memBuf.UpdateFlags(key, tikv.SetKeyLocked, tikv.DelNeedCheckExists, valExists)
 	}
 	txn.lockedCnt += len(keys) - skipedLockKeys
 	return nil
 }
 
+func (txn *KVTxn) resetPrimary() {
+	txn.committer.primaryKey = nil
+	txn.committer.ttlManager.reset()
+}
+
+func (txn *KVTxn) selectPrimaryForPessimisticLock(sortedKeys [][]byte) {
+	if txn.aggressiveLockingContext != nil {
+		lastPrimaryKey := txn.aggressiveLockingContext.lastPrimaryKey
+		if txn.aggressiveLockingContext.lastPrimaryKey != nil {
+			foundIdx := sort.Search(len(sortedKeys), func(i int) bool {
+				return bytes.Compare(sortedKeys[i], lastPrimaryKey) >= 0
+			})
+			if foundIdx < len(sortedKeys) && bytes.Equal(sortedKeys[foundIdx], lastPrimaryKey) {
+				// The last selected primary is included in the current list of keys we are going to lock. Select it
+				// as the primary.
+				txn.committer.primaryKey = sortedKeys[foundIdx]
+				txn.aggressiveLockingContext.assignedPrimaryKey = true
+				txn.aggressiveLockingContext.primaryKey = sortedKeys[foundIdx]
+			} else {
+				txn.committer.primaryKey = sortedKeys[0]
+				txn.aggressiveLockingContext.assignedPrimaryKey = true
+				txn.aggressiveLockingContext.primaryKey = sortedKeys[0]
+			}
+		} else {
+			txn.committer.primaryKey = sortedKeys[0]
+			txn.aggressiveLockingContext.assignedPrimaryKey = true
+			txn.aggressiveLockingContext.primaryKey = sortedKeys[0]
+		}
+	} else {
+		txn.committer.primaryKey = sortedKeys[0]
+	}
+}
+
+type aggressiveLockingContext struct {
+	lastRetryUnnecessaryLocks map[string]tempLockBufferEntry
+	lastPrimaryKey            []byte
+	lastAttemptStartTime      time.Time
+
+	currentLockedKeys       map[string]tempLockBufferEntry
```

Comment:
How to track the `tempLockBufferEntry` related using the new memory statistic mechanism for a transaction? Is it necessary? 
@ekexium PTAL

Reply:
Currently we only track the memory in memdb via callback, which should be the most significant part.
In the future we could consider introducing an individual tracker for each transaction which contains child trakcers for memdb, `aggressiveLockingContext` , snapshot cache, or anything necessary.